server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -96,6 +99,7 @@
     private final StartRecoveryRequest request;
     private final int chunkSizeInBytes;
     private final RecoveryTargetHandler recoveryTarget;
+    private final AtomicReference<Consumer<Exception>> onCancel = new AtomicReference<>();
```

Comment:
I don't like this. Can we rather add callback support to `CancellableThreads` this would make a lot of things much simpler?

Reply:
++